runtime/src/crypto.rs

```diff
@@ -0,0 +1,128 @@
+// Licensed under the Apache-2.0 license
+
+use caliptra_common::{crypto::Ecc384KeyPair, keyids::KEY_ID_TMP};
+use caliptra_drivers::{
+    hmac384_kdf, Array4x12, Ecc384PrivKeyOut, Ecc384PubKey, Hmac384Data, Hmac384Key, Hmac384Tag,
+    KeyId, KeyReadArgs, KeyUsage, KeyWriteArgs,
+};
+use caliptra_error::CaliptraResult;
+use zerocopy::AsBytes;
+use zeroize::Zeroize;
+
+use crate::Drivers;
+
+pub enum Crypto {}
+
+impl Crypto {
+    // Calculate an HMAC-384 KDF
+    fn hmac384_kdf(
+        drivers: &mut Drivers,
+        key: KeyId,
+        label: &[u8],
+        context: Option<&[u8]>,
+        output: KeyId,
+    ) -> CaliptraResult<()> {
+        hmac384_kdf(
+            &mut drivers.hmac384,
+            KeyReadArgs::new(key).into(),
+            label,
+            context,
+            &mut drivers.trng,
+            KeyWriteArgs::new(
+                output,
+                KeyUsage::default()
+                    .set_hmac_key_en()
+                    .set_ecc_key_gen_seed_en(),
+            )
+            .into(),
+        )
+    }
+
+    // Generate an ECC keypair
+    fn ecc384_key_gen(
+        drivers: &mut Drivers,
+        input: KeyId,
+        label: &[u8],
+        priv_key: KeyId,
+    ) -> CaliptraResult<Ecc384KeyPair> {
+        Crypto::hmac384_kdf(drivers, input, label, None, KEY_ID_TMP)?;
+
+        let key_out = Ecc384PrivKeyOut::Key(KeyWriteArgs::new(
+            priv_key,
+            KeyUsage::default().set_ecc_private_key_en(),
+        ));
+
+        let pub_key = drivers.ecc384.key_pair(
+            &KeyReadArgs::new(KEY_ID_TMP).into(),
+            &Array4x12::default(),
+            &mut drivers.trng,
+            key_out,
+        );
+
+        if KEY_ID_TMP != priv_key {
+            drivers.key_vault.erase_key(KEY_ID_TMP)?;
+        }
+
+        Ok(Ecc384KeyPair {
+            priv_key,
+            pub_key: pub_key?,
+        })
+    }
+
+    // "Hash" the data in the provided KV slot by HMACing it with an empty slice.
+    // This mechanism is necessary because the hardware does not directly support
+    // hashing data in KV slots.
+    pub fn hmac384_hash(drivers: &mut Drivers, input: KeyId, output: KeyId) -> CaliptraResult<()> {
+        drivers.hmac384.hmac(
+            &KeyReadArgs::new(input).into(),
+            &Hmac384Data::Slice(&[]),
+            &mut drivers.trng,
+            KeyWriteArgs::new(
+                output,
+                KeyUsage::default()
+                    .set_hmac_key_en()
+                    .set_ecc_key_gen_seed_en(),
+            )
+            .into(),
+        )
+    }
+
+    // Perform an "HMAC" with a key from KV by first using it to derive an
+    // ECC keypair, then hashing the public key coordinates into an HMAC key.
+    // This roundabout mechanism is necessary because the hardware does not
+    // directly support exposing an HMAC computed with key material from KV.
```

Comment:
probably worth a note that this is only secure because the generated public key is kept secret. Just so future contributors know the public key shouldn't be exposed.

Reply:
Added a note.